File: backend/app/api/v1/endpoints/notifications.py
from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, get_db
from app.models.user import User
from app.schemas.notification import NotificationCountRead, NotificationRead
from app.services import notification_service

router = APIRouter(prefix="/notifications", tags=["notifications"])


@router.get("", response_model=dict)
async def list_notifications(
    offset: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=200),
    unread_only: bool = Query(False),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    items, total = await notification_service.list_notifications(
        db, user.id, offset=offset, limit=limit, unread_only=unread_only,
    )
    return {
        "items": [NotificationRead.model_validate(n) for n in items],
        "total": total,
        "offset": offset,
        "limit": limit,
    }


@router.get("/unread-count", response_model=NotificationCountRead)
async def get_unread_count(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    count = await notification_service.get_unread_count(db, user.id)
    return NotificationCountRead(unread_count=count)


@router.post("/{notification_id}/read", status_code=status.HTTP_204_NO_CONTENT)
async def mark_notification_read(
    notification_id: UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    success = await notification_service.mark_as_read(db, notification_id, user.id)
    if not success:
        raise HTTPException(status_code=404, detail="Notification not found")


@router.post("/read-all", status_code=status.HTTP_200_OK)
async def mark_all_read(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    count = await notification_service.mark_all_read(db, user.id)
    return {"marked_read": count}

File: frontend/src/api/notifications.ts
import apiClient from './client'
import type { PaginatedResponse } from '@/types/api'

export interface Notification {
  id: string
  user_id: string
  event_type: string
  title: string
  body: string | null
  entity_type: string | null
  entity_id: string | null
  data: Record<string, unknown>
  is_read: boolean
  read_at: string | null
  created_at: string
}

export async function listNotifications(params?: {
  offset?: number
  limit?: number
  unread_only?: boolean
}): Promise<PaginatedResponse<Notification>> {
  const { data } = await apiClient.get<PaginatedResponse<Notification>>('/notifications', { params })
  return data
}

export async function getUnreadCount(): Promise<number> {
  const { data } = await apiClient.get<{ unread_count: number }>('/notifications/unread-count')
  return data.unread_count
}

export async function markAsRead(notificationId: string): Promise<void> {
  await apiClient.post(`/notifications/${notificationId}/read`)
}

export async function markAllRead(): Promise<{ marked_read: number }> {
  const { data } = await apiClient.post<{ marked_read: number }>('/notifications/read-all')
  return data
}
